src/Dhall/TypeCheck.hs

```diff
@@ -154,11 +155,29 @@ typeWithA tpa = loop
                     then return ()
                     else Left (TypeError ctx e (AnnotMismatch a0 nf_A0 nf_A1))
             Nothing -> return ()
-        let a1 = Dhall.Core.normalize a0
-        let a2 = Dhall.Core.shift 1 (V x 0) a1
-        let b1 = Dhall.Core.subst (V x 0) a2 b0
-        let b2 = Dhall.Core.shift (-1) (V x 0) b1
-        loop ctx b2
+
+        t <- loop ctx _A1
+
+        -- The catch-all branch will actually do the correct thing for
+        -- Type as well. However, the Type branch will be much faster
+        -- if a0 is large in some sense: we type-check it once here,
+        -- and then don't need to again at the use sites; this can be
+        -- a very significant speed-up - see #412.
+        case Dhall.Core.normalize t of
+          Const Type -> do
```

Comment:
Have you tried doing this fast-path for all expressions (i.e. not just terms)?

Reply:
Oh, never mind.  I see from your comment on the original issue that you don't want to do this for types because then it breaks support for type synonyms

Note that this implies that if we ever turn on true dependent types (i.e. functions from terms to types) then we would have to eliminate this fast path.  Not that I plan on introducing dependent type support any time soon, but it's worth adding a comment to this line:

https://github.com/dhall-lang/dhall-haskell/blob/6f626c96bcb79ea7d1b421487fad69392ad7e8c0/src/Dhall/TypeCheck.hs#L66

... explaining that if we turn on dependent types then we need to turn off this fast path to avoid a type-checking bug